File: QuickSort.java
public class Solution {

	public void sort(int[] A) {

		quickSort(A, 0, A.length -1);

		return;

	}

	void quickSort(int[] A, int start, int end) {

		if (start >= end) {
			return;
		}
		
		// 每次选择以中间这个数作为pivot
		// 目的是把所有小于这个pivot的放在左边
		// 所有大于这个pivot的放在右边
		int mid = (start + end)/2;
		int pivot = A[mid];
		
		// 这种交换可以通过双指针实现
		int left = start, right = end;

		while (left <= right) {
			
                        // 找到第一个大于等于pivot的数
			while (left <=right && A[left] < pivot) {
				left ++;
			}
                        // 找到第一个小于等于pivot的数
			while (left <= right && A[right] > pivot) {
				right --;
			}

			// swap
			if (left <= right) {
				int temp = A[left];
				A[left] = A[right];
				A[right] = left;

				left ++;
				right --;
			}
		}
		
		// 然后分别quick sort 左边和右边
		quickSort(A, start, left);
		quickSort(A, right, end);
	}

}

File: quick_sort.py
class Solution:
	
	def sort(self, A):
		self.quick_sort(A, 0, len(A) - 1)
		
# start - the start index
# end - the end index
    def quickSort(self, A, start, end):
		if start >= end:
			return
		left, right = start, end
		pivot = A[(left + right) // 2]
		while left + 1 < right:
			while left + 1 < right and A[left] < pivot:
				left += 1
			while left + 1 < right and A[right] > pivot:
				right -= 1
			if left + 1 < right:
				A[left], A[right] = A[right], A[left]
				left += 1
				right -=1
		
		self.quickSort(A, start, left)
		self.quickSort(A, right, end)
